src/quic/streams.cc

```diff
@@ -1,61 +1,1052 @@
+  if (!source || is_destroyed() || !is_writable()) return;
+  DCHECK_NULL(outbound_);
+  outbound_ = std::make_unique<Outbound>(this, std::move(source));
+  session_->ResumeStream(id());
+}
+
+void Stream::EntryRead(size_t amount) {
+  // Tells us that amount bytes were read from inbound_
+  // We use this as a signal to extend the flow control
+  // window to receive more bytes.
+  if (is_destroyed()) return;
+  if (session_) session_->ExtendStreamOffset(id(), amount);
+}
+
+int Stream::DoPull(bob::Next<ngtcp2_vec> next,
+                   int options,
+                   ngtcp2_vec* data,
+                   size_t count,
+                   size_t max_count_hint) {
+  if (is_destroyed() || is_eos()) {
+    std::move(next)(bob::Status::STATUS_EOS, nullptr, 0, [](int) {});
+    return bob::Status::STATUS_EOS;
+  }
+
+  // If an outbound source has not yet been attached, block until one is
+  // available. When AttachOutboundSource() is called the stream will be
+  // resumed. Note that when we say "block" here we don't mean it in the
+  // traditional "block the thread" sense. Instead, this will inform the
+  // Session to not try to send any more data from this stream until there
+  // is a source attached.
+  if (outbound_ == nullptr) {
+    std::move(next)(bob::Status::STATUS_BLOCK, nullptr, 0, [](size_t len) {});
+    return bob::Status::STATUS_BLOCK;
+  }
+
+  return outbound_->Pull(std::move(next), options, data, count, max_count_hint);
+}
+
+void Stream::BeginHeaders(HeadersKind kind) {
+  if (is_destroyed()) return;
+  headers_length_ = 0;
+  headers_.clear();
+  headers_kind_ = kind;
+}
+
+bool Stream::AddHeader(const Header& header) {
+  if (is_destroyed()) return false;
+  size_t len = header.length();
+  if (!session_->application().CanAddHeader(
+          headers_.size(), headers_length_, len)) {
+    return false;
+  }
+
+  headers_length_ += len;
+
+  auto& state = BindingData::Get(env());
+
+  const auto push = [&](auto raw) {
+    Local<Value> value;
+    if (UNLIKELY(!raw.ToLocal(&value))) return false;
+    headers_.push_back(value);
+    return true;
+  };
+
+  return push(header.GetName(&state)) && push(header.GetValue(&state));
+}
+
+void Stream::Acknowledge(size_t datalen) {
+  if (is_destroyed() || outbound_ == nullptr) return;
+
+  // ngtcp2 guarantees that offset must always be greater than the previously
+  // received offset.
+  DCHECK_GE(datalen, STAT_GET(Stats, max_offset_ack));
+  STAT_SET(Stats, max_offset_ack, datalen);
+
+  // // Consumes the given number of bytes in the buffer.
+  outbound_->Acknowledge(datalen);
+}
+
+void Stream::Commit(size_t datalen) {
+  if (is_destroyed() || outbound_ == nullptr) return;
+  outbound_->Commit(datalen);
 }
 
-void Stream::Acknowledge(size_t datalen) {}
-void Stream::Blocked() {}
-void Stream::Commit(size_t datalen) {}
-void Stream::End() {}
-void Stream::Destroy(QuicError error) {}
+void Stream::EndWritable() {
+  if (is_destroyed() || !is_writable()) return;
```

Comment:
Question: Should we add a CHECK instead of these if statements and make sure that this function is not executed with these properties?

Reply:
Yeah, I've got a note to revisit all of these as part of the performance pass once I've got the basic impl ready.